core/lib/spree/core/number_generator.rb

```diff
@@ -0,0 +1,44 @@
+module Spree
+  module Core
+    class NumberGenerator < Module
+      RAND           = Random.new
+      BASE           = 10
+      DEFAULT_LENGTH = 9
+      NUMBERS        = (0..9).to_a.freeze
+      LETTERS        = ('A'..'Z').to_a.freeze
+
+      def initialize(options)
+        @prefix     = options.fetch(:prefix)
+        @length     = options.fetch(:length, DEFAULT_LENGTH)
+        @candidates = NUMBERS + (options[:letters] ? LETTERS : [])
+      end
+
+      def included(host)
+        host.validates(:number, presence: true, uniqueness: true)
+
+        generator = method(:generate_permalink)
+        host.before_validation do |instance|
+          instance.number ||= generator.call(host)
+        end
+      end
+
+      private
+
+      def generate_permalink(host)
+        length = @length
+
+        loop do
+          candidate = new_candidate(length)
+          return candidate unless host.exists?(number: candidate)
+
+          # If over half of all possible options are taken add another digit.
+          length += 1 if host.count > Rational(BASE**length, 2)
+        end
+      end
+
+      def new_candidate(length)
+        @prefix + Array.new(length) { @candidates.fetch(RAND.rand(@candidates.length)) }.join
```

Comment:
Without that constant its less testable, since I do not have a hook to "inject" my test random generator. On my next pass where we replace the monkey patch configuration "feature" I'll look into this idea. See: https://github.com/spree/spree/pull/6413#discussion_r31182826

But I think your idea is better, will see how I can kill all mutations.